### ISSUE TITLE
Binding calls twice setter of properties it is bound to

### ISSUE THREAD

**Issue by kociak** [nion-software/nionutils#2]
Hi

I am experiencing a problem with binding when using a property (instead of a Model), inside a plug-in. This problem seems similar in declarative UI and the old style UI.
Binding a the text of a line_edit to a property triggers a double call to the property setter. The problem appears with or without the use of a `Converter`. The problem appears wether or not I trigger an event in the setter.

Minimal example is given below (note that in my case, the object holding the property is defined in a different file.

`Class Handler:
    def __init__(self,instrument):
        self.instrument=instrument
        self.text = Model.PropertyModel(10)
        self.wavelength=self.instrument.wavelength
        self.wavelength = self.instrument.simpleproperty
        self.enabled = False
        self.wavelength_converter=Converter.PhysicalValueToStringConverter("nm", 1E9)
    def init_push(self,widget):
        self.instrument.init(1, 8)
      
class View:
    def __init__(self):
        ui = Declarative.DeclarativeUI()
   
        self.ui_view=ui.create_row(ui.create_push_button(text="Init", name="init_pb", on_clicked="init_push"),ui.create_stretch(),ui.create_line_edit(text="@binding(instrument.wavelength,converter=wavelength_converter)"),ui.create_line_edit(text="@binding(text.value,converter=wavelength_converter)"))
        self.ui_view=ui.create_row(self.ui_view,ui.create_check_box(text="Enable", name="enable_cb", checked="enabled"))

#### This property is defined in the instrument class

@property
	def wavelength(self) -> float:
		return self.__wavelength

	@wavelength.setter
	def wavelength(self, value: float) -> None:
		self.__wavelength=value
		logging.info("inside setter, value is {}".format(value))
		if (self.last_wavelength==value):
			logging.info("double call, value is {}".format(value))
		self.last_wavelength=value
`

**Comment by cmeyer** [comment]
I cannot reproduce this problem. You might try putting the following code inside of `wavelength.setter` to see where it is being called from:

```
import traceback
traceback.print_stack()
```

I modified the UI Demo to test this by changing the `LineEdits.py` file to the code below. Do you see any place this is different from your code?

It's a bit complicated, but you can modify the UI demo to use the code below.

You either need to have `pyqt` installed or `nionui-tool` installed to run the demo. If you're already running Nion Swift from your Python environment, just install `nionui-tool` with `conda install -c nion nionui-tool`. Then run the demo using:

`nionui nionui_app.nionui_examples.ui_demo`

Then you can modify `nionui/nionui_app/nionui_examples/ui_demo/LineEdits.py` within your Python environment's site_packages folder.

Once you have modified the file and run it, select the 'Line Edit' page in the chooser menu item at the top of the window. Then edit the field on the right and hit return. You'll see the new value only gets printed once.

```
from nion.utils import Converter
from nion.utils import Model

class Handler:

    def __init__(self):
        self.color_line_edit = None
        self.abcd = Model.PropertyModel(value=1.6e-8)

        self.converter = Converter.PhysicalValueToStringConverter("nm", 1E9)

        def value_changed(value):
            print(value)

        self.abcd.on_value_changed = value_changed

    def color_updated(self, widget, text):
        print(f"Color changed: {widget} {text}")

    def return_pressed(self, widget):
        self.color_line_edit.select_all()
        return False  # required so that editing_finished event is triggered

    def key_pressed(self, widget, key):
        if key.text == "*":
            return True
        return False


def construct_ui(ui):
    field_label = ui.create_label(text="Favorite Color?")
    field_line_edit = ui.create_line_edit(placeholder_text="Color", name="color_line_edit", on_editing_finished="color_updated", on_return_pressed="return_pressed", on_key_pressed="key_pressed")
    second_edit = ui.create_line_edit(text="@binding(abcd.value, converter=converter)")
    return ui.create_row(ui.create_spacing(8), field_label, ui.create_spacing(8), field_line_edit, ui.create_spacing(8), second_edit, ui.create_spacing(8))
```

**Comment by kociak** [comment]
thank for the help. 
There is a clear difference between your code and mine.
I am not using a `Model` but I am using a property (in the python sense).
I can reproduce your code, no problem, it works.
However if B is a property of object A, where A is defined in another file but has an instance which is an attribute of the `Handler`, then I get an error.
A second weirdness is that the call to the property (here, `wavelength`) is trigerred again when the `field_line_edit` lose focus, wether or not we have stroke the enter key.

Here is the stack call:

```
File "/anaconda/envs/nionswift/bin/Nion Swift.app/Contents/MacOS/../Resources/bootstrap.py", line 24, in bootstrap_dispatch
    return getattr(object, method_name)(*args)

File "/anaconda/envs/nionswift/lib/python3.6/site-packages/nion/ui/QtUserInterface.py", line 1184, in returnPressed
    return self.on_return_pressed()
```
File "/anaconda/envs/nionswift/lib/python3.6/site-packages/nion/ui/UserInterface.py", line 1320, in handle_return_pressed
    return self.on_return_pressed()

File "/anaconda/envs/nionswift/lib/python3.6/site-packages/nion/ui/UserInterface.py", line 1418, in return_pressed
    self.__binding.update_source(text)

File "/anaconda/envs/nionswift/lib/python3.6/site-packages/nion/utils/Binding.py", line 112, in update_source
    self.source_setter(converted_value)

File "/anaconda/envs/nionswift/lib/python3.6/site-packages/nion/utils/Binding.py", line 246, in <lambda>
    self.source_setter = lambda value: setattr(self.source, self.__property_name, value)

File "/Users/mathieukociak/PycharmProjects/orsay/swift-spectro/nionswift_plugin/spectro/SimpleSpectro.py", line 167, in wavelength
    traceback.print_stack()
```

I hope this helps!

**Comment by cmeyer** [comment]
Reference code. Be sure to call `self.property_changed_event.fire("abcd")` in the property setter.

```
from nion.utils import Converter
from nion.utils import Event

class Handler:

    def __init__(self):
        self.color_line_edit = None
        self.__abcd = 1.6e-8
        self.property_changed_event = Event.Event()

        self.converter = Converter.PhysicalValueToStringConverter("nm", 1E9)

    @property
    def abcd(self):
        return self.__abcd

    @abcd.setter
    def abcd(self, value):
        self.__abcd = value
        self.property_changed_event.fire("abcd")
        print(value)

    def color_updated(self, widget, text):
        print(f"Color changed: {widget} {text}")

    def return_pressed(self, widget):
        self.color_line_edit.select_all()
        return False  # required so that editing_finished event is triggered

    def key_pressed(self, widget, key):
        if key.text == "*":
            return True
        return False


def construct_ui(ui):
    field_label = ui.create_label(text="Favorite Color?")
    field_line_edit = ui.create_line_edit(placeholder_text="Color", name="color_line_edit", on_editing_finished="color_updated", on_return_pressed="return_pressed", on_key_pressed="key_pressed")
    second_edit = ui.create_line_edit(text="@binding(abcd, converter=converter)")
    return ui.create_row(ui.create_spacing(8), field_label, ui.create_spacing(8), field_line_edit, ui.create_spacing(8), second_edit, ui.create_spacing(8))
```

**Comment by cmeyer** [comment]
This fix is now deployed as nionui 0.3.19 on both PyPI and Anaconda cloud.

**Comment by kociak** [comment]
Works like a charm now
